src/inference.jl

```diff
@@ -24,25 +24,25 @@ end
 
 @kwdef struct TransformerLayerWeights
     # weights for rmsnorms
-    rms_att_weight::Vector{Float32} # (dim,)
-    rms_ffn_weight::Vector{Float32} # (dim,)
+    rms_att_weight::AbstractVector{Float32} # (dim,) # usually Float32
+    rms_ffn_weight::AbstractVector{Float32} # (dim,) # usually Float32
     # weights for matmuls
-    wq::Matrix # (dim, dim)
-    wk::Matrix # (dim, dim)
-    wv::Matrix # (dim, dim)
-    wo::Matrix # (dim, dim)
+    wq::AbstractMatrix # (dim, dim)
+    wk::AbstractMatrix # (dim, dim)
+    wv::AbstractMatrix # (dim, dim) # different quantization usually
+    wo::AbstractMatrix # (dim, dim)
     # weights for ffn
-    w1::Matrix # (dim, hidden_dim)
-    w2::Matrix # (hidden_dim, dim)
-    w3::Matrix # (dim, hidden_dim)
+    w1::AbstractMatrix # (dim, hidden_dim)
+    w2::AbstractMatrix # (hidden_dim, dim) # different quantization usually
+    w3::AbstractMatrix # (dim, hidden_dim)
 end
 
 @kwdef struct TransformerWeights
-    token_embedding_table::Matrix # (dim, vocab_size)
-    layers::Vector{TransformerLayerWeights}
+    token_embedding_table::AbstractMatrix # (dim, vocab_size)
+    layers::Vector{TransformerLayerWeights} # NTuple{TransformerLayerWeights} # but not every layer is the same type!
     # final rmsnorm
-    rms_final_weight::Vector{Float32} # (dim,)
-    output_weight::Matrix # (dim, vocab_size)
+    rms_final_weight::AbstractVector # (dim,) # Float32
+    output_weight::AbstractMatrix # (dim, vocab_size)
 end
```

Comment:
Since `TransformerLayerWeights` is a parametric type, it might be useful here to also do

```julia
@kwdef struct TransformerWeights{T<:TransformerLayerWeights}
    token_embedding_table::AbstractMatrix # (dim, vocab_size)
    layers::Vector{T} # NTuple{TransformerLayerWeights} # but not every layer is the same type!
    # final rmsnorm
    rms_final_weight::AbstractVector # (dim,) # Float32
    output_weight::AbstractMatrix # (dim, vocab_size)
end
...

Reply:
I guess for maximalising performance layers could be `Tuple`, so every type it holds is available when we iterate over the `TransformerLayerWeights`.

Reply:
Yeah, if @cafaxo is into it then I think tuples would be appropriate here, though it introduces a different consideration that the unwrapped tuples may be union types rather than type-stable: https://discourse.julialang.org/t/parametric-type-in-an-variable-length-tuple/9109/10?u=cpfiffer

Not sure if that's likely to happen here, that's probably a benchmarking thing for sure.